sys/dev/gpio/acpi_gpiobus.c

```diff
@@ -0,0 +1,304 @@
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHORS AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <sys/types.h>
+#include <sys/bus.h>
+#include <sys/kernel.h>
+#include <sys/module.h>
+#include <sys/gpio.h>
+
+#include <contrib/dev/acpica/include/acpi.h>
+#include <dev/acpica/acpivar.h>
+
+#include <dev/gpio/gpiobusvar.h>
+
+struct acpi_gpiobus_softc {
+	struct gpiobus_softc	super_sc;
+	ACPI_CONNECTION_INFO	handler_info;
+};
+
+struct acpi_gpiobus_ctx {
+	struct gpiobus_softc	*sc;
+	ACPI_HANDLE		dev_handle;
+};
+
+static ACPI_STATUS
+acpi_gpiobus_enumerate_res(ACPI_RESOURCE *res, void *context)
+{
+	ACPI_RESOURCE_GPIO *gpio_res = &res->Data.Gpio;
+	struct acpi_gpiobus_ctx *ctx = context;
+	struct gpiobus_softc *super_sc = ctx->sc;
+	ACPI_HANDLE handle;
+	uint32_t flags, i;
+
+	if (res->Type != ACPI_RESOURCE_TYPE_GPIO)
+		return (AE_OK);
+
+	if (ACPI_FAILURE(AcpiGetHandle(ACPI_ROOT_OBJECT,
+	    gpio_res->ResourceSource.StringPtr, &handle)) ||
+	    handle != ctx->dev_handle)
+		return (AE_OK);
+
+	/* Figure out pin flags */
+	flags = 0;
+
+	if (gpio_res->ConnectionType == ACPI_RESOURCE_GPIO_TYPE_INT)
+		switch (gpio_res->Polarity) {
+		case ACPI_ACTIVE_HIGH:
+			flags = gpio_res->Triggering == ACPI_LEVEL_SENSITIVE ?
+			    GPIO_INTR_LEVEL_HIGH : GPIO_INTR_EDGE_RISING;
+			break;
+		case ACPI_ACTIVE_LOW:
+			flags = gpio_res->Triggering == ACPI_LEVEL_SENSITIVE ?
+			    GPIO_INTR_LEVEL_LOW : GPIO_INTR_EDGE_FALLING;
+			break;
+		case ACPI_ACTIVE_BOTH:
+			flags = GPIO_INTR_EDGE_BOTH;
+			break;
+		}
+	else
+		switch (gpio_res->IoRestriction) {
+		case ACPI_IO_RESTRICT_INPUT:
+			flags = GPIO_PIN_INPUT;
+			break;
+		case ACPI_IO_RESTRICT_OUTPUT:
+			flags = GPIO_PIN_OUTPUT;
+			break;
+		}
+
+	switch (gpio_res->PinConfig) {
+	case ACPI_PIN_CONFIG_PULLUP:
+		flags |= GPIO_PIN_PULLUP;
+		break;
+	case ACPI_PIN_CONFIG_PULLDOWN:
+		flags |= GPIO_PIN_PULLDOWN;
+		break;
+	}
```

Comment:
There is also gpio_res->Shareable field to process. It should be mapped to new GPIO_INTR_SHARED flag in sys/gpio.h

Reply:
Oh, nice catch.

Would we need to change the interrupt handling code in gpioc to accommodate the flag?
But then again, I think this attribute only means "Can this be accessed/is this used by other ACPI devices". So in that case, it shouldn't affect interrupt creation.
Hm, I'm not sure.